hathor/transaction/storage/tx_allow_scope.py

```diff
@@ -0,0 +1,68 @@
+# Copyright 2023 Hathor Labs
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from contextlib import contextmanager
+from enum import Flag, auto
+from typing import TYPE_CHECKING, Generator
+
+from hathor.conf import HathorSettings
+from hathor.transaction.base_transaction import BaseTransaction
+
+if TYPE_CHECKING:
+    from hathor.transaction.storage import TransactionStorage  # noqa: F401
+
+
+settings = HathorSettings()
+
+
+class TxAllowScope(Flag):
+    """ This enum is used internally to mark which "type" of transactions to allow the database to read/write
+
+    In this context "type" it means validation level, the supported "types" are enumerated in this class and for the
+    purpose of filtering it can be any combination of the supported types.
+    """
+    VALID = auto()
+    PARTIAL = auto()
+    INVALID = auto()
+
+    def is_allowed(self, tx: BaseTransaction) -> bool:
+        """True means it is allowed to be used in the storage (as argument or as return), False means not allowed."""
+        tx_meta = tx.get_metadata()
+        # XXX: initial/partial/invalid/fully_connected never overlap and cover all possible validation states
+        #      see hathor.transaction.transaction_metadata.ValidationState for more details
+        validation = tx_meta.validation
```

Comment:
Should we add `assert validation in list_with_all_possible_states`?

Reply:
In this case `list_with_all_possible_states = list(ValidationState)`. Asserting that in itself is pretty redundant with the whole Enum class.

What maybe wouldn't be redundant is `assert list(ValidationState) = [ValidationState.INITIAL, ValidationState.BASIC, ...]`, which would basically assert what the expected enum variants are. I'm not sure if it makes sense to do this here though. Maybe it would make more sense to have a simple test for that which could point to the known code locations that make certain assumptions based on the enum variants. I'm not too sure how much this helps though.

What do you think?

Reply:
`assert list(ValidationState) == [ValidationState.INITIAL, ...]` is exactly what I meant. But I guess we can pass it. It would be great to have a mechanism that guarantees we have covered all possible cases otherwise a linter would fail or an exception would be raised.

Reply:
I'm actually tweaking this a little bit because of some edge cases (when using full validations we temporarily save transactions with INITIAL state). I'll add a test to assert that assumption (in the XXX comment) and the state list.

Reply:
Done. I've added those tests.

Reply:
Great! Thanks!